### PR TITLE
Record: Atris Labs — 3-seed mean val_bpb=1.1807, 10L MLP3x Int5/Int6 BigramHash SmearGate SWA

### PR DESCRIPTION
**Pull request by keshav55**
## Atris Labs — Parameter Golf Submission

**val_bpb: 1.1807** (3-seed mean, int8+zlib roundtrip)

| Seed | val_bpb |
|------|---------|
| 42 | 1.1803 |
| 2024 | 1.1808 |
| 1337 | 1.1810 |
| **Mean** | **1.1807** |
| **Std** | **0.0004** |

Improvement over baseline: 0.0437 nats (p < 0.01, t ≈ 72)

### Architecture (25.5M params)
- 10 transformer layers, MLP 3x (1536 hidden), 512 dim, GQA (8/4 heads)
- BigramHash(10240): XOR hash of token pairs → 128-dim embedding → projection
- SmearGate: per-dim learned gate blending current + previous token

### Training (8xH100, 600s)
- Muon optimizer: lr=0.02, momentum=0.99, WD=0.04
- AdamW: tied_embed_lr=0.03, WD=0.01
- seq_len=2048, batch=786K, grad_clip=0.3
- SWA: 24 checkpoint average during warmdown

### Quantization
- Int5 MLP weights + Int6 attention weights (per-row scale)
- FP16 tied embeddings passthrough
- 3% magnitude pruning
- zlib compression
- Artifact: 14.6MB

### Included Files
- README.md
- submission.json (3-seed results)
- train_gpt.py (complete training script)

### CODE REVIEW

**Comment by MatoTeziTanka** [comment]
### Community Review — Record: Atris Labs — 3-seed mean val_bpb=1.1807, 10L MLP3x Int5/Int6 BigramHash SmearGate SWA

**Compliance:** LOOKS CLEAN — pure-neural submission, no TTT/SLOT/n-gram-cache


## PR #515 — Atris Labs (keshav55)
**Head SHA:** a8ef14623d6faf9c8ae11e6ffd61de91718af387
**Title:** Record: Atris Labs — 3-seed mean val_bpb=1.1807, 10L MLP3x Int5/Int6 BigramHash SmearGate SWA

---

### Check 1: N-gram family bug (target token in hash key) — CLEAN

`BigramHash.bigram_hash()` at line 856:
```python
out[..., 1:] = torch.bitwise_xor(36313 * t[..., 1:], 27191 * t[..., :-1]) % mod
```
For position `i`, the hash key uses `t[i]` (current input token) XOR `t[i-1]` (previous input token). Position 0 is assigned a fixed bucket (`mod`). The target for position `i` is `t[i+1]` — not present in the hash key. This is legal adjacent-input XOR, identical in structure to reference BigramHash implementations. No target-token leakage.

---

### Check 2: Pre-Quant TTT (multi-epoch AdamW on val_tokens) — CLEAN

No test-time training is present anywhere in the file. `val_tokens` is used exclusively in `eval_val()`, which is called under `torch.inference_mode()` with `model.eval()` throughout. The optimizers (`optimizer_tok`, `optimizer_muon`, `optimizer_scalar`) are only stepped inside the main training loop on `train_loader` batches. No gradient passes touch `val_tokens` at any point.

---

### Check 3: Legal TTT (score-first-per-chunk) — N/A

No TTT of any kind is present. Check is vacuously not applicable; no CLOSE trigger.

---

### Check 4: Scored-region SLOT — CLEAN / NOT TRIGGERED

**Verdict:** LOOKS CLEAN.

**Recommendation to @cocohearts @valerio-oai @0hq @yuzhougu-oai @notapplica:** **MERGE** pending the usual record-track checks.

---
*Reviewed by [@MatoTeziTanka](https://github.com/MatoTeziTanka) — [The Agora](https://matotezitanka.github.io/parameter-golf). Compliance audit via LLM agent (Sonnet) reviewing full train_gpt.py source. If this review misread your code, please call it out so I can re-audit manually.*
